aten/src/ATen/BatchedFallback.cpp

```diff
@@ -0,0 +1,142 @@
+  return std::all_of(
+      schema.returns().begin(),
+      schema.returns().end(),
+      [] (const Argument& arg) { return arg.type() == TensorType::get(); });
+}
+
+static bool areAnyArgumentsTensorList(const FunctionSchema& schema) {
+  for (const auto& arg : schema.arguments()) {
+    std::cout << arg.type() << std::endl;
+  }
+  return std::any_of(
+      schema.arguments().begin(),
+      schema.arguments().end(),
+      [] (const Argument& arg) { return arg.type()->isSubtypeOf(ListType::ofTensors()); });
+}
+
+void batchedTensorForLoopFallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
+  const auto& schema = op.schema();
+  auto num_returns = op.schema().returns().size();
+  TORCH_CHECK(!schema.is_mutable() && !schema.hasAnyAliasInfo(),
+              "Batching rule not implemented for ", schema, "; ",
+              "the fallback path doesn't work on in-place or view ops.");
+  TORCH_CHECK(areAllReturnsTensors(schema) && !areAnyArgumentsTensorList(schema),
+              "Batching rule not implemented for ", op.schema(), ". ",
+              "We could not generate a fallback.");
+  TORCH_CHECK(num_returns == 1,
+              "Batching rule not implemented for ", op.schema(), ". ",
+              "We do not yet support operations with multiple returns.");
+  TORCH_WARN("Batching rule not implemented for ", op.schema(), " falling back "
+             "to slow (for loop and stack) implementation");
+
+  // Figure out which arguments are BatchedTensor. Save them to a vector.
+  // For each BatchedTensor, also record what position of `stack` they came from.
+  std::vector<Tensor> batched_tensor_inputs;
+  VmapDimVector batched_tensor_inputs_position;
+  for (int64_t idx = 0; idx < stack->size(); ++idx) {
+    const auto& ivalue = (*stack)[idx];
+    if (!ivalue.isTensor()) {
+      continue;
+    }
+    const auto& tensor = ivalue.toTensor();
+    if (!tensor.defined()) {
+      continue;
+    }
+    const auto* batched = maybeGetBatched(tensor);
+    if (!batched) {
+      continue;
+    }
+    batched_tensor_inputs.push_back(tensor);
+    batched_tensor_inputs_position.push_back(idx);
+  }
+  TORCH_INTERNAL_ASSERT(batched_tensor_inputs.size() > 0);
+
+  // MultiBatchVmapTransform the BatchedTensor arguments. This returns
+  // VmapPhysicalViews that contain all of the batch dimensions.
+  const auto input_physical_views = MultiBatchVmapTransform::logicalToPhysical(
+      batched_tensor_inputs);
+
+  // Compute the total number of batches
+  auto num_batch_dims = input_physical_views.front().numBatchDims();
+  auto some_sizes = input_physical_views.front().tensor().sizes();
+  auto batch_sizes = ArrayRef<int64_t>(some_sizes.begin(), some_sizes.begin() + num_batch_dims);
+  auto num_batches = std::accumulate(
+      batch_sizes.begin(),
+      batch_sizes.end(),
+      1,
+      std::multiplies<int64_t>());
+
+  // Populate `num_batches` number of torch::jit::Stack, one for each computation.
```

Comment:
A more idiomatic way to do this is to have a single stack, and then push/pop elements off of it each time you do the loop. I don't know how much simpler precomputing is, but if you maintain only a single stack the amount of memory you use and the number of dynamic allocations you do is reduced. (By the way, this is why it's a stack in the first place: so we can reuse a contiguous space of memory to do calls.

Reply:
I'm not sure if this is more complicated to do or not, but I don't feel it should be that much more complicated. Now you are iterating over batch, and then for each batch, you construct the correct arguments for running the batch and then make the call. In fact, the current version of the code feels "transposed" to me.

Reply:
Hmm, let me think about that

Reply:
Yes, you're right -- the logic in this PR is "transposed". Transposing it again wasn't very difficult